[WASM-Function-References] Assertion failure when a function returns (ref <idx>) to JS

takikawa opened the pull request:
#### 0c946f31a45171fb24601ae6e31a95ac874858e4
<pre>
[WASM-Function-References] Assertion failure when a function returns (ref &lt;idx&gt;) to JS
<a href="https://bugs.webkit.org/show_bug.cgi?id=243265">https://bugs.webkit.org/show_bug.cgi?id=243265</a>

Reviewed by Justin Michaud.

This patch fixes some JS interaction cases in typed function reference
support. In particular, if a `(ref $t)` type is used at a JS&lt;-&gt;Wasm
boundary, assertion failures were possible in some cases.

When typed funcrefs were turned on, it was also possible for existing
interactions to fail. For example, for `funcref` uses because it is
normalized to `(ref null func)`.

The patch fixes this by generally treating Ref/RefNull type kinds the
same way that Externref/Funcref type kinds are in WasmToJS.cpp.

This patch also adds checking for uses of struct type indices to raise
an error when they are used in JS interactions through Ref/RefNull
types. When the Wasm GC JS API is more well-defined, these error cases
can likely be removed.

* ../../../JSTests/wasm/function-references/bug243265.js: Added.
(module):
(async exportRefIndexResultFunc):
(async refIndexArgToJS):
* ../../../JSTests/wasm/gc/structs.js:
* ../../../Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::callInformationFor const):
* ../../../Source/JavaScriptCore/wasm/WasmExceptionType.h:
* ../../../Source/JavaScriptCore/wasm/WasmGlobal.cpp:
(JSC::Wasm::Global::set):
* ../../../Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
* ../../../Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::handleBadGCTypeIndexUse):
(JSC::Wasm::wasmToJS):
(JSC::Wasm::emitThrowWasmToJSException):

Canonical link: <a href="https://commits.webkit.org/253734@main">https://commits.webkit.org/253734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/apple/WebKit/commit/f3c8dc53d04f494090164e7c57b7395c016319a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9222 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/104455 "The change is no longer eligible for processing.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/49925 "The change is no longer eligible for processing.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27407 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/104455 "The change is no longer eligible for processing.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/49925 "The change is no longer eligible for processing.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102331 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14641 "Found 1 new test failure: fast/dom/StyleSheet/detached-style-pi.xhtml (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89676 "Found 1 new API test failure: TestWebKitAPI.FullscreenVideoTextRecognition.DoNotAnalyzeVideoAfterExitingFullscreen (failure)") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/104455 "The change is no longer eligible for processing.") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14435 "Found 1 new test failure: imported/w3c/web-platform-tests/css/cssom/stylesheet-same-origin.sub.html (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7657 "Found 1 new test failure: fast/dom/StyleSheet/detached-style-pi.xhtml (failure)") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/49285 "The change is no longer eligible for processing.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84360 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7744 "Found 1 new test failure: fast/dom/StyleSheet/detached-style-pi.xhtml (failure)") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106812 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26438 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19281 "Found 2 new test failures: fast/dom/StyleSheet/detached-style-pi.xhtml imported/w3c/web-platform-tests/html/capability-delegation/delegate-fullscreen-request-subframe-same-origin.https.tentative.html (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/106812 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26800 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85878 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106812 "The change is no longer eligible for processing.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28732 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6420 "Found 2 new test failures: fast/dom/StyleSheet/detached-style-pi.xhtml pdf/pdf-plugin-printing.html (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20173 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26378 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31578 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26198 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -419,7 +413,7 @@ Expected<MacroAssemblerCodeRef<WasmEntryPtrTag>, BindingFailure> wasmToJS(VM& vm
             break;
         }
         default:  {
-            if (Wasm::isFuncref(returnType) || Wasm::isExternref(returnType) || Wasm::isI31ref(returnType))
```
I think we should make sure that we don't export internal WASM struct or array references until that part of the spec is ironed out. 

Otherwise, LGTM

takikawa replied:
> I think we should make sure that we don't export internal WASM struct or array references until that part of the spec is ironed out.

This took a little bit more code, since it required adding additional error checking masm code in `WasmToJS.cpp` and `JSToWasm.cpp` (as you want the check at run-time, to avoid erroring when create the Wasm-JS stubs). I modeled the check after the old check that was used for disallowed I64 exports to JS.

The latest push has the new checks and some additional tests that exercise them. Also some minor fixes, like making `emitThrowWasmToJSException` explicitly pass the third argument to the operation call instead of implicitly assuming it's in the right register.

justinmichaud commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -492,6 +510,10 @@ void emitThrowWasmToJSException(CCallHelpers& jit, GPRReg wasmInstance, Wasm::Ex
     jit.copyCalleeSavesToEntryFrameCalleeSavesBuffer(GPRInfo::argumentGPR0);
     jit.move(GPRInfo::callFrameRegister, GPRInfo::argumentGPR0);
+
+    if (wasmInstance != GPRInfo::argumentGPR2)
```
Isn't this pinned? I am confused now

justinmichaud replied:
Hmm, what is the difference between this and wasmContextInstanceGPR?

takikawa replied:
Since the operation that's being called (`operationWasmToJSException`) takes the instance as the third argument, the instance pointer has to be moved into GPR2 or be there already.

Previously all callers of this function just passed in GPR2 (and it would fail in weird ways if you passed in other registers), so I just changed it here to work if you passed a different one in (such as the pinned register).

justinmichaud reviewed (approved):
r=me with an answer to the comment about the instanceGPR

justinmichaud commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -53,6 +53,25 @@ static void materializeImportJSCell(JIT& jit, unsigned importIndex, GPRReg resul
+    jit.prepareCallOperation(vm);
+    jit.emitPutToCallFrameHeader(GPRInfo::argumentGPR1, CallFrameSlot::callee);
+
+    emitThrowWasmToJSException(jit, GPRInfo::argumentGPR2, Wasm::ExceptionType::InvalidGCTypeUse);
```
Why do we assume this is GPR2 and not the pinned instance register?

takikawa replied:
We could use the pinned register here (as in the code in `JSToWasm.cpp`), but since it has to be moved into GPR2 anyway in `emitThrowWasmToJSException` there isn't much difference. (this was based on the old I64 checking code which did the same)

justinmichaud reviewed (approved):
r=me, ping me when you need merge-queue

takikawa committed:
[WASM-Function-References] Assertion failure when a function returns (ref <idx>) to JS
https://bugs.webkit.org/show_bug.cgi?id=243265

Reviewed by Justin Michaud.

This patch fixes some JS interaction cases in typed function reference
support. In particular, if a `(ref $t)` type is used at a JS<->Wasm
boundary, assertion failures were possible in some cases.

When typed funcrefs were turned on, it was also possible for existing
interactions to fail. For example, for `funcref` uses because it is
normalized to `(ref null func)`.

The patch fixes this by generally treating Ref/RefNull type kinds the
same way that Externref/Funcref type kinds are in WasmToJS.cpp.

This patch also adds checking for uses of struct type indices to raise
an error when they are used in JS interactions through Ref/RefNull
types. When the Wasm GC JS API is more well-defined, these error cases
can likely be removed.

* ../../../JSTests/wasm/function-references/bug243265.js: Added.
(module):
(async exportRefIndexResultFunc):
(async refIndexArgToJS):
* ../../../JSTests/wasm/gc/structs.js:
* ../../../Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::callInformationFor const):
* ../../../Source/JavaScriptCore/wasm/WasmExceptionType.h:
* ../../../Source/JavaScriptCore/wasm/WasmGlobal.cpp:
(JSC::Wasm::Global::set):
* ../../../Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
* ../../../Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::handleBadGCTypeIndexUse):
(JSC::Wasm::wasmToJS):
(JSC::Wasm::emitThrowWasmToJSException):

Canonical link: https://commits.webkit.org/253734@main
